crates/circuits/mod-builder/src/core_chip.rs

```diff
@@ -242,59 +299,101 @@ where
-        self.name.clone()
-    }
+    type RecordLayout = AdapterCoreLayout<FieldExpressionMetadata>;
+    type RecordMut<'a> = (A::RecordMut<'a>, FieldExpressionCoreRecordMut<'a>);
 
-    fn execute(
+    fn execute<'buf, RA>(
         &mut self,
         state: VmStateMut<TracingMemory<F>, CTX>,
         instruction: &Instruction<F>,
-        trace: &mut [F],
-        trace_offset: &mut usize,
-        width: usize,
-    ) -> Result<()> {
-        let row_slice = &mut trace[*trace_offset..*trace_offset + width];
-        let (adapter_row, core_row) = row_slice.split_at_mut(A::WIDTH);
+        arena: &'buf mut RA,
+    ) -> Result<()>
+    where
+        RA: RecordArena<'buf, Self::RecordLayout, Self::RecordMut<'buf>>,
+    {
+        let core_record_metadata = FieldExpressionMetadata {
+            total_input_limbs: self.num_inputs() * self.expr.canonical_num_limbs(),
+        };
+
+        let (mut adapter_record, mut core_record) = arena.alloc(AdapterCoreLayout::with_metadata(
+            A::WIDTH,
+            core_record_metadata,
+        ));
 
-        A::start(*state.pc, state.memory, adapter_row);
+        A::start(*state.pc, state.memory, &mut adapter_record);
 
         let data: DynArray<_> = self
             .adapter
-            .read(state.memory, instruction, adapter_row)
+            .read(state.memory, instruction, &mut adapter_record)
             .into();
 
-        let (writes, inputs, flags) = run_field_expression(self, &data, instruction);
+        core_record.fill_from_execution_data(
+            instruction.opcode.local_opcode_idx(self.offset) as u8,
+            &data.0,
+        );
 
-        // TODO(arayi): Should move this to fill_trace_row
-        self.expr
-            .generate_subrow((self.range_checker.as_ref(), inputs, flags), core_row);
+        let (writes, _, _) = run_field_expression(self, &data.0, core_record.inner.opcode as usize);
 
-        self.adapter
-            .write(state.memory, instruction, adapter_row, &writes.into());
+        self.adapter.write(
+            state.memory,
+            instruction,
+            &writes.into(),
+            &mut adapter_record,
+        );
 
         *state.pc = state.pc.wrapping_add(DEFAULT_PC_STEP);
-        *trace_offset += width;
         Ok(())
     }
 
-    fn fill_trace_row(&self, mem_helper: &MemoryAuxColsFactory<F>, row: &mut [F]) {
-        let (adapter_row, _) = row.split_at_mut(A::WIDTH);
-        self.adapter.fill_trace_row(mem_helper, (), adapter_row);
+    fn get_opcode_name(&self, _opcode: usize) -> String {
+        self.name.clone()
+    }
+}
+
+impl<F, CTX, A> TraceFiller<F, CTX> for FieldExpressionStep<A>
+where
+    F: PrimeField32 + Send + Sync + Clone,
+    A: 'static + AdapterTraceFiller<F, CTX>,
+{
+    fn fill_trace_row(&self, mem_helper: &MemoryAuxColsFactory<F>, row_slice: &mut [F]) {
+        // NOTE[teokitan]: This is where GPU acceleration should happen in the future
```

Comment:
remove: `TraceFiller` is for cpu only, this entire function and more will be replaced by a separate kernel